indexer/index.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2016 The Vulcan Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indexer
+
+import (
+	"regexp"
+	"sync"
+)
+
+// Index can resolve matchers to fully qualified metric names.
+type Index struct {
+	entries map[string]map[string]string
+	m       sync.RWMutex
```

Comment:
I favored the non-pointer because its zero value is usable. All `func (i *Index) ...` methods must be done by pointer receiver though, which is my default.

These two blog posts are what shaped my opinion on this.
https://dave.cheney.net/2013/01/19/what-is-the-zero-value-and-why-is-it-useful
https://dave.cheney.net/2016/03/19/should-methods-be-declared-on-t-or-t